### PR TITLE
Rework on animals code

### LINKED ISSUE #525: Familiarization decays faster than Classic
I don't have any numerical data on this yet, but it seems like familiarity level decays rapidly after the 24 hour period is up and animals are familiarizable again. I seem to recall in Classic that they wouldn't lose their current familiarity level until they haven't been fed for 48 hours.

**Meta Info**
 - TFC Version: 0.25.3 build 87

### LINKED ISSUE #526: Shearing any sheep color other then white will give incorrect wool texture
**Describe the bug**
Make sure to include:
1. What did you expect to happen? get regular tfc wool texture
2. What actually happened instead (i.e. what was the bug) get unknown texture wool

**To Reproduce**
Attention to detail is important!
1. Do '...' download any version of tng
2. Then do '...' shear a sheep that is not white

**Meta Info**
 - TFC Version: 87

pretty sure this is related to TFC sheep class EntitySheepTFC importing net.minecraft.entity.passive.EntitySheep;
which has the following statement
@Nullable
    protected ResourceLocation getLootTable() {
        if (this.getSheared()) {
            return LootTableList.ENTITIES_SHEEP;
        } else {
            switch(this.getFleeceColor()) {
            case WHITE:
            default:
                return LootTableList.ENTITIES_SHEEP_WHITE;
            case ORANGE:
                return LootTableList.ENTITIES_SHEEP_ORANGE;
            case MAGENTA:
                return LootTableList.ENTITIES_SHEEP_MAGENTA;
            case LIGHT_BLUE:
                return LootTableList.ENTITIES_SHEEP_LIGHT_BLUE;
            case YELLOW:
                return LootTableList.ENTITIES_SHEEP_YELLOW;
            case LIME:
                return LootTableList.ENTITIES_SHEEP_LIME;
            case PINK:
                return LootTableList.ENTITIES_SHEEP_PINK;
            case GRAY:
                return LootTableList.ENTITIES_SHEEP_GRAY;
            case SILVER:
                return LootTableList.ENTITIES_SHEEP_SILVER;
            case CYAN:
                return LootTableList.ENTITIES_SHEEP_CYAN;
            case PURPLE:
                return LootTableList.ENTITIES_SHEEP_PURPLE;
            case BLUE:
                return LootTableList.ENTITIES_SHEEP_BLUE;
            case BROWN:
                return LootTableList.ENTITIES_SHEEP_BROWN;
            case GREEN:
                return LootTableList.ENTITIES_SHEEP_GREEN;
            case RED:
                return LootTableList.ENTITIES_SHEEP_RED;
            case BLACK:
                return LootTableList.ENTITIES_SHEEP_BLACK;
            }
        }
    }

### LINKED ISSUE #640: Baby bears too powerful; should be scaled down.
Nerf baby bears. Presently, they have 60 health and are extremely fast and aggressive. If anything, they are probably more powerful than adults. Given the attempt at realism, it's not to be expected that bear cubs are aggressive (1) and (2) even adult bears are not necessarily aggressive and probably should behave more like vanilla polar bears than endermen. 

In the very least, bear cubs should flee when struck if they become aggressive, but should not attack unprovoked. They also should not have more health than every single monster in the game. 

TFC Version: MC1.12.2-0.27.3.103

### LINKED ISSUE #685: Animal's offspring don't inherit mother's familarity?
**Describe the bug**
Make sure to include:
1. What did you expect to happen?
Bred animal offspring (pigs in this instance) to inherit some of the mother's familiarity.
2. What actually happened instead (i.e. what was the bug)
None was inherited.

**Meta Info**
 - TFC Version: 0.X.X.107

### PR DESCRIPTION
**Pull request by DisasterMoo**
I was chatting with Abdyman about another idea for an addon. A bridge between TFC and [Animania](https://www.curseforge.com/minecraft/mc-mods/animania). So, before I can work on it, there was a thing bothering me in animals code, that is, I should've used interfaces instead of abstract classes for implementing most animals.

So this PR is about it. I'm moving things from `EntityAnimalTFC` to `IAnimalTFC` for easier animal implementations in the future (ie: possibility of using base vanilla animals + implementing this interface so TFC an work with it).

Needs to be addressed yet:
- [x] Group Spawning
- [x] ~~Some form of indication for old animals~~(left that out for the time being. old aging only indication should be texture or maybe a TOP integration)

Also, just found out I fixed #525 
Fixed #685 
Fixed #526 by only dropping white wool regarless of fleece color.
Closes #640 by making baby bears not aggressive unless hit

### CODE REVIEW

**Comment by DisasterMoo** [comment]
Lastly, there was that thing about genetic selection, but i'm gonna leave that out from this PR.. For now, sheeps only drops white wool, and are only white.

**Comment by DisasterMoo** [comment]
Merged Abdyman's into this PR
Supersedes #529

**Comment by Gaelmare** [comment]
What's the status on this?

**Comment by DisasterMoo** [comment]
On hold until we get the priority list done.

File: src/main/java/net/dries007/tfc/ConfigTFC.java [hunk 1]
```diff
@@ -307,6 +307,20 @@ public static void onConfigChangedEvent(ConfigChangedEvent.OnConfigChangedEvent
         @Config.Comment("The default length of a month (in days) when a new world is started. This can be changed in existing worlds via the /timetfc command.")
         @Config.LangKey("config." + MOD_ID + ".general.defaultMonthLength")
         public int defaultMonthLength = 8;
+
+        @Config.Comment("Should animals became old and die?")
+        @Config.LangKey("config." + MOD_ID + ".general.enableAnimalAging")
+        public boolean enableAnimalAging = true;
```
Comment (alcatrazEscapee) [inline src/main/java/net/dries007/tfc/ConfigTFC.java]: What does being old do to animals - just prevent breeding + reproducing, and have a chance of dying every day?

File: src/main/java/net/dries007/tfc/ConfigTFC.java [hunk 2]
```diff
@@ -307,6 +307,20 @@ public static void onConfigChangedEvent(ConfigChangedEvent.OnConfigChangedEvent
...
+
+        @Config.Comment("How long until animals became old (in factor to adulthood)?")
+        @Config.RangeDouble(min = 1, max = 50)
+        @Config.LangKey("config." + MOD_ID + ".general.factorAnimalAging")
+        public double factorAnimalAging = 3;
+
+        @Config.Comment("Chance of animal dying (checked every in-game day) after it became old")
+        @Config.RangeDouble(min = 0, max = 1)
+        @Config.LangKey("config." + MOD_ID + ".general.chanceAnimalDeath")
+        public double chanceAnimalDeath = 0.05;
```
Comment (alcatrazEscapee) [inline src/main/java/net/dries007/tfc/ConfigTFC.java]: I think this should be set to 0 by default, I'm not entirely convinced it's a feature that won't just lead to excessive respawning and/or death of animals. Is there any visual indication of getting old?

Comment (Abdyman) [reply src/main/java/net/dries007/tfc/ConfigTFC.java]: i think old animals should no longer be able to breed, but remove the option of them dying. atm theres no visual indication of them getting old as far as im aware

File: src/main/java/net/dries007/tfc/ConfigTFC.java (hunk 1 above)
Comment (DisasterMoo) [reply src/main/java/net/dries007/tfc/ConfigTFC.java]: Also prevents produce (milk, egg, wool)

File: src/main/java/net/dries007/tfc/ConfigTFC.java (hunk 2 above)
Comment (DisasterMoo) [reply src/main/java/net/dries007/tfc/ConfigTFC.java]: Yes, no indication. I will make their textures location update via code and duplicate in the assets, so, in time, someone better than me can do something about it 😛

File: src/main/java/net/dries007/tfc/objects/entity/animal/EntityDeerTFC.java [hunk 3]
```diff
@@ -78,25 +95,15 @@ public long gestationDays()
...
-        return (float) value;
+        return DAYS_TO_ADULTHOOD;
     }
 
     @Override
-    public Age getAge()
+    public boolean isFood(ItemStack stack)
     {
-        return CalendarTFC.PLAYER_TIME.getTotalDays() >= this.getBirthDay() + DAYS_TO_ADULTHOOD ? Age.ADULT : Age.CHILD;
+        return stack.getItem() == ItemsTFC.SALT;
```
[the remark this answers is not in the document]
Comment (DisasterMoo) [reply src/main/java/net/dries007/tfc/objects/entity/animal/EntityDeerTFC.java]: Strange, isn't it? That came from classic. But deers aren't tameable, so, it is only used for enticing iirc.

**Pull request merged**